### PR TITLE
Add package dependency network graph to the blueprint

### PR DESCRIPTION
**Pull request by vedhav**
Hey, @insightsengineering/nest-core-dev 👋
What do you think about adding a package dependency network graph in our product map vignette of the blueprint?
It was super helpful for me to understand how different packages connect to each other.

This adds the network diagram to explore the package dependencies of the teal framework to the Technical Blueprint's Product Map.

It seems like more dependencies are added to suggests but only `visNetwork` and `config` are the new suggests and the rest of them are already indirect dependencies of `teal`.

P.S I used the four primary colors from the NEST logo for the node groups 😉

### CODE REVIEW

File: vignettes/blueprint/config.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        - teal.widgets
+        - teal.reporter
+        - teal
+    Modules:
+      color: "#f24b4b"
+      packages:
+        - teal.osprey
+        - teal.goshawk
+        - teal.modules.clinical
+        - teal.modules.helios
```
Comment (vedhav) [inline vignettes/blueprint/config.yml:24]: Question: Do we need `teal.modules.helios` here or do we add it here after the public release is made? If we only use public repos we don't need GitHub token to access the `staged_dependencies.yaml` and the pkgdown site will build without any issues.

**Comment by pawelru** [comment]
Note that this is quite lengthy process to complete - are we good with that? Especially that it's in CI that got re-triggered quite frequently?

I feel that there could be much faster way of getting that information using `pkgdepends` that relies on metadata in DESC file and not cloning the whole repository. 

Also, I really doubt such visualisation on the NEST level will give a lot of _value_. There are quite a few issues about deps already so this is nothing new I feel. If you want to enhance it more - I would prefer to think more repo-wise than project-wise. I think it could be more actionable.

**Comment by vedhav** [comment]
Agree that `staged_dependencies.yaml` should not be used for this. Especially when we [intend to remove them](https://github.com/openpharma/mmrm/pull/383) after the cran release is made (I'm surprised our CI will work without this file)
I do see some value from this. The mermaid graph in the product map is perfect in conveying the product architecture of teal however this does not convey the inter-dependencies between different teal core packages and it is easy to create a false mental model that `teal.code`, `teal.data`, and `teal.*` are imported into `teal` and they don't have much to do with each other. Someone has to go through each of the package dependency individually to understand this connection. This felt like the easiest way for showing this.

**Comment by gogonzo** [comment]
> Especially when we https://github.com/openpharma/mmrm/pull/383 after the cran release is made (I'm surprised our CI will work without this file)

I can't imagine work without `staged.dependencies` when testing feature branches. `@main` is covered by [r-universe](https://pharmaverse.r-universe.dev/) and releasing packages to CRAN doesn't add any value to the cicd process.

**Comment by pawelru** [comment]
> Agree that `staged_dependencies.yaml` should not be used for this. Especially when we [intend to remove them](https://github.com/openpharma/mmrm/pull/383) after the cran release is made (I'm surprised our CI will work without this file)

Honestly, I am quite surprised seeing this and agree with what @gogonzo said. The staged-deps and CRAN should be independent. CRAN is a deployment repository and staged-deps allows to simulate mono-repo for multiple-repo development purposes. I'm not a big fan of this but it seems that sometimes it's needed - hence: we don't want to remove it (or more precisely: we are not ready to remove it).

> I do see some value from this. The mermaid graph in the product map is perfect in conveying the product architecture of teal however this does not convey the inter-dependencies between different teal core packages and it is easy to create a false mental model that teal.code, teal.data, and teal.* are imported into teal and they don't have much to do with each other. Someone has to go through each of the package dependency individually to understand this connection. This felt like the easiest way for showing this.

OK - I understand. Then let's continue discussing. We can do this two ways:

1) Modify the existing chart
This means hard-coding dependency map and therefore it's likely that the information become outdated soon... Maybe we can try to be more _conceptual_ than _correct_ - e.g. just make a connection between teal.code and teal and don't go into the details on what kind of dependency it is etc. Just an idea how to address that problem... From what I can see now - that was the intention.

2) Use some other functionality?
Staged-deps was designed for something other. It's might be too slow (too powerful) just to present a dependency graph.
Moreover, (like I said earlier) I am not a big fan of staged-deps and sooner or later I will bring it's need into the discussion. Right now it's too early because our products are being too dependent between each other. But we have made a significant progress on that regard. Let's see in the future.
I mentioned `pkgdepends` earlier. Please find a snippet below to get the relevant data for getting the whole map. You can move forward from this.
``` r
x <- pkgdepends::new_pkg_deps(c("insightsengineering/teal", "insightsengineering/teal.data", "insightsengineering/teal.code", "insightsengineering/teal.slice"))
x$solve()
#> 
#> ✔ Updated metadata database: 4.88 MB in 8 files.
#> 
#> ℹ R 4.3 aarch64-apple-darwin20 packages are missing from Bioconductor
#> ℹ Updating metadata database
#> ✔ Updating metadata database ... done
#> 
y <- x$get_resolution()[, c("package", "direct", "deps")]

y
#> # A data frame: 213 × 3
#>    package     direct deps          
#>    <chr>       <lgl>  <list>        
#>  1 teal        TRUE   <tbl [50 × 5]>
#>  2 teal.data   TRUE   <tbl [18 × 5]>
#>  3 teal.code   TRUE   <tbl [21 × 5]>
#>  4 teal.slice  TRUE   <tbl [44 × 5]>
#>  5 BH          FALSE  <tbl [0 × 5]> 
#>  6 MASS        FALSE  <tbl [10 × 5]>
#>  7 Matrix      FALSE  <tbl [14 × 5]>
#>  8 R.cache     FALSE  <tbl [6 × 5]> 
#>  9 R.methodsS3 FALSE  <tbl [3 × 5]> 
#> 10 R.oo        FALSE  <tbl [5 × 5]> 
#> # ℹ 203 more rows
y[1, "deps"][[1]]
#> # A data frame: 50 × 5
#>    ref            type    package        op    version     
#>    <chr>          <chr>   <chr>          <chr> <chr>       
#>  1 R              Depends R              ">="  "4.0"       
#>  2 shiny          Depends shiny          ">="  "1.7.0"     
#>  3 teal.data      Depends teal.data      ">="  "0.3.0.9017"
#>  4 teal.slice     Depends teal.slice     ">="  "0.4.0.9027"
#>  5 teal.transform Depends teal.transform ">="  "0.4.0.9010"
#>  6 checkmate      Imports checkmate      ">="  "2.1.0"     
#>  7 jsonlite       Imports jsonlite       ""    ""          
#>  8 lifecycle      Imports lifecycle      ">="  "0.2.0"     
#>  9 logger         Imports logger         ">="  "0.2.0"     
#> 10 magrittr       Imports magrittr       ">="  "1.5"       
#> # ℹ 40 more rows
```

<sup>Created on 2023-12-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by vedhav** [comment]
Thank you @pawelru for suggesting a simpler way of fetching the dependencies.
Circling back on this as we're cleaning the vignettes before the CRAN release. What are your thoughts on showing the Dependencies of the `teal` packages like this.
Screenshot 2024-01-24 at 9 37 24 AM

**Comment by chlebowa** [comment]
Personally, I don't find this particularly easy to read. I would have to stop thinking about text and focus on this graph for more than a minute to comprehend it OR give it a glance and ignore it and continue reading. Again, this is me personally, maybe other people may have an easier time with it.

**Review by gogonzo — CHANGES_REQUESTED** [review]
Six extra dependencies in the Suggests just to make dependencies graph - I don't feel good with this. 
Also, I don't think that user/developer learns a lot about the product by knowing what depends on what. This is redundant information IMO. We should picture how these relationships look like (and we do in the vignettes).

**Comment by chlebowa** [comment]
> `dplyr` is anyway going to be installed as it's imported by `teal.code`

It most certainly is not.
```
Type: Package
Package: teal.code
Title: Code Storage and Execution Class for 'teal' Applications
Version: 0.5.0.9000
Date: 2024-01-11
Authors@R (parsed):
    * Dawid Kaledkowski <dawid.kaledkowski@roche.com> [aut, cre]
    * Aleksander Chlebowski <aleksander.chlebowski@contractors.roche.com> [aut]
    * Marcin Kosinski <marcin.kosinski.mk1@roche.com> [aut]
    * Pawel Rucki <pawel.rucki@roche.com> [aut]
    * Nikolas Burkoff <nikolas.burkoff@roche.com> [aut]
    * Mahmoud Hallal <mahmoud.hallal@roche.com> [aut]
    * Maciej Nasinski <maciej.nasinski@contractors.roche.com> [aut]
    * Konrad Pagacz <konrad.pagacz@contractors.roche.com> [aut]
    * Junlue Zhao <zhaoj88@gene.com> [aut]
    * Chendi Liao <chendi.liao@roche.com> [rev]
    * Dony Unardi <unardid@gene.com> [rev]
    * F. Hoffmann-La Roche AG [cph, fnd]
Description: Introduction of 'qenv' S4 class, that facilitates
    code execution and reproducibility in 'teal' applications.
License: Apache License 2.0
URL: https://insightsengineering.github.io/teal.code/,
    https://github.com/insightsengineering/teal.code
BugReports:
    https://github.com/insightsengineering/teal.code/issues
Depends:
    methods,
    R (>= 4.0)
Imports:
    checkmate (>= 2.1.0),
    grDevices,
    lifecycle (>= 0.2.0),
    rlang (>= 1.1.0)
Suggests:
    cli (>= 3.4.0),
    knitr (>= 1.42),
    magrittr (>= 1.5),
    rmarkdown (>= 2.19),
    shiny (>= 1.6.0),
    testthat (>= 3.1.5)
VignetteBuilder:
    knitr
RdMacros:
    lifecycle
Config/Needs/verdepcheck: mllg/checkmate, r-lib/lifecycle,
    r-lib/rlang, rstudio/shiny, r-lib/styler, r-lib/cli, yihui/knitr,
    tidyverse/magrittr, rstudio/rmarkdown, r-lib/testthat
Config/Needs/website: insightsengineering/nesttemplate
Encoding: UTF-8
Language: en-US
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.2.3
Collate:
  (...)
```

**Comment by vedhav** [comment]
I meant to say `teal.slice`. My bad.

**Comment by gogonzo** [comment]
> Updated the diagram to show the edges with "imports" and "depends", Flipped the direction so it makes sense and Improved the layout.

I still don't find it useful. I think graph doesn't have any value added for the users. It just visually unattractive map of our products. Map doesn't say anything but "these are our products and this is how we complicated relationships between them". What we should do (and did already):
- we should just explain how to make data for teal and teal.data is the right package
- how include reporter in teal and that teal.reporter is needed for this
- etc.

No one except posit cares that `dplyr` depends on `rlang`.

File: vignettes/blueprint/product_map.Rmd [hunk 2]
```diff
@@ -80,3 +81,53 @@ By breaking down `teal` features, modules, and calculations into dedicated R pac
...
+
+teal_packages <- c(
+  "teal.logger", "teal.code", "teal.data", "teal.slice",
+  "teal.widgets", "teal.reporter", "teal", "teal.transform"
+)
+teal_deps <- new_pkg_deps(
+  paste0("insightsengineering/", teal_packages)
+)
+teal_deps$solve()
+resolved_teal_deps <- teal_deps$get_resolution()[, c("package", "direct", "deps")] |>
```
[the remark this answers is not in the document]
Comment (vedhav) [reply vignettes/blueprint/product_map.Rmd:103]: 🤦🏽 I just got so used to it that I didn't realise that we never use it in our packages. Let me change it to magrittr pipe

**Comment by m7pr** [comment]
Hey, sorry for the late input. I also think the diagram is rather more helpful for internal developers and it's not a landing page for end users so I don't think it gives any value for users

**Comment by vedhav** [comment]
@gogonzo Fair enough. I'll close this PR in that case. The value add (if any) does not justify the addition of new dependencies.

**Pull request closed without merging**